src/Parallafka/MessageCommitter.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Parallafka.KafkaConsumer;
+
+namespace Parallafka
+{
+    internal class MessageCommitter<TKey, TValue>
+    {
+        private readonly IKafkaConsumer<TKey, TValue> _consumer;
+        private readonly CommitState<TKey, TValue> _commitState;
+        private readonly ILogger _logger;
+        private readonly CancellationToken _stopToken;
+        private readonly CancellationTokenSource _stopTimer;
+        private readonly Task _commitTimer;
+        private readonly Dictionary<int, CommitPartitionState> _committedStates;
+
+        public MessageCommitter(
+            IKafkaConsumer<TKey, TValue> consumer,
+            CommitState<TKey, TValue> commitState,
+            ILogger logger,
+            TimeSpan timerDelay,
+            CancellationToken stopToken)
+        {
+            this._committedStates = new();
+            this._consumer = consumer;
+            _commitState = commitState;
+            this._logger = logger;
+            this._stopToken = stopToken;
+            this.Completion = new TaskCompletionSource().Task;
+            this._stopTimer = new();
+            this._commitTimer = Task.Run(() => this.CommitOnTimer(timerDelay));
+        }
+
+        /// <summary>
+        /// Indicates that the message committer should complete all possible commits.
+        /// Await <see cref="Completion"/> to know when the committer is completed.
+        /// </summary>
+        public void Complete()
+        {
+            this._stopTimer.Cancel();
+            Completion = Task.WhenAll(this._commitTimer, GetAndCommitAnyMessages());
```

Comment:
Should this be sequential?
```
Task WaitForCompletionAsync()
{
    await this._commitTimer;
    await GetAndCommitAnyMessages();
}
Completion = WaitForCompletionAsync();
```

My race condition senses were tingling but not clear

Reply:
No: It doesn't matter what order they finish. 
The commitTimer  calls GetAndCommitAnyMessages and this await also calls GetAndCommitAnyMessages.
There is a "race condition" if both were calling GetAndCommitAnyMessages at the same time, yes.  But the race condition is safe: GetAndCommitAnyMessages is thread-safe because GetMessagesToCommit and GetMessageToCommit are threadsafe.
But it's a good thought to be concerned that they COULD cause a problem.

Reply:
Having said all that - you can eliminate this concern by just having the timer do one more GetAndCommitAnyMessages call before exiting the loop.

This, combined with your idea to have only 1 committer thread - namely the timer - gave me some ideas.
Move the timer into its own class (maybe CommitState?) and we can simplify some of the locks and eliminate read locks at the very least and potentially write locks.

Reply:
See: https://github.com/Snagajob/parallafka-for-dotnet/pull/1